### PR TITLE
FFmpegImage: Reduce quality a tiny bit

### PR DESCRIPTION
**Pull request by fritsch**
While reading the forum I have seen that a user copied his Thumbnails
directory to a PC and "optimized" the jpegs. From this I checked what
we do in code and have seen that we use the "maximum" possible quality
when storing a lossy jpeg format. This results in very large jpegs
which from quality POV hardly to distinguish at all from a little
compress image. I chose 4 as a sensible value.

Size comparisons:
53K Mai 18 19:16 1b005dd9.jpg
53K Mai 18 19:40 1b005dd9-2.jpg
43K Mai 18 19:40 1b005dd9-3.jpg
30K Mai 18 19:40 1b005dd9-4.jpg
28K Mai 18 19:41 1b005dd9-5.jpg
26K Mai 18 19:41 1b005dd9-6.jpg
22K Mai 18 19:41 1b005dd9-7.jpg
21K Mai 18 19:41 1b005dd9-8.jpg

## What is the effect on users?
Thumbnails will look normally but only use 50% of the size.

## Screenshots (if appropriate):

qscale=1 (highest quality) towards qscale=8 (lowest quality tested)
Original (1):
1b005dd9
2:
1b005dd9-2
3:
1b005dd9-3
4:
1b005dd9-4
5:
1b005dd9-5
6:
1b005dd9-6
7:
1b005dd9-7
8:
1b005dd9-8

### CODE REVIEW

**Comment by fritsch** [comment]
Curious if there are better ideas / feedback from some people with a certain library, concerning size and visual regression (if there is).

**Comment by L2501** [comment]
Maybe also for PNG add -pred mixed?

**Comment by rmrector** [comment]
Nice. I fiddled with putting this into AS.xml some time ago, a patch gave my poor Vero with 16GB storage a few extra months of life. I was thinking 4 or 5 would be good for the default value. The results were still notably better looking than streaming service apps on my TV, though they also use a lower resolution.

**Comment by CastagnaIT** [comment]
just a consideration compression usually gives its worst in cases with panoramas where details do not reach the furthest point, testing with close-ups will not make much difference

**Comment by zag2me** [comment]
Very nice improvement!
Certainly I have issues with the the total size of the thumbnail cache on my shield.
4 seems like a good comprimise.

**Comment by fritsch** [comment]
> just a consideration compression usually gives its worst in cases with panoramas where details do not reach the furthest point, testing with close-ups will not make much difference

I gave the ffmpeg command above. Mind posting an example? Q1 vs Q4?

**Comment by da-anda** [comment]
with jpeg I wouldn't go too far "down" (quality wise), since especially red areas tend to get noticeable artifacts. So it's good that your test picture contained a lot of red 👍 
Personally, for me 4 would be the absolute min quality, because you can already see some slight block artifacts in red background areas (see light ray above the guy in the back on the left). I would probably tend to 3, no objections to 4 though.

**Comment by da-anda** [comment]
I based my reply on your test images, which clearly show the block artifacts in the mentioned area (at least I can see them). I also speak from experience working a lot with jpeg files over the last 10-15 years, but I have no experience with ffmpeg's image converter. But in Photoshop or ImageMagick you most of the time end up with quality factor of 70-80 (of 100) when red areas especially with sharp edges are involved (design team will complain if you go with lower quality settings in those cases). With no red areas involved you can go down to 55-60 for a still somewhat acceptable quality.

So if we are fine that certain edge cases might show clear artifacts in red areas, we could even go with q5 I suppose. If we aim for good quality pictures that likely covers all edge cases, q3 might be better. But as a compromise the suggested q4 would still be fine with me based on your test image. Not sure I can find/come up with an image that highlights the compression artifacts of the jpeg algo in red areas, I only had those cross my desk several times already.

**Comment by fritsch** [comment]
Got it. Thanks for explanation. I entirely misread your statement. Yes I agree and that was why I chose 4 :-) sorry for the noise.

**Comment by da-anda** [comment]
np, although I was a bit confused by your reply :) Btw, nice to see you around fritsch

**Comment by zag2me** [comment]
My tests show very little visual quality loss on higher resolution source images from metadata sites. Honestly there isnt much quality loss from testing with higher resolution images.

4 is a great level I think, but it can also go a few steps lower and still maintain good quality.
**Quality 1**

**Quality 2**

**Quality 3**

**Quality 4**

**Quality 5**

**Quality 6**

**Quality 7**

**Quality 8**


Screenshot 2022-05-20 095056

**Comment by fritsch** [comment]
Anything left here? Changing the default?

From a project POV I would not make this change too late in the cycle, e.g. not after the first beta - then we get most feedback. Honestly I think no one will realize, but that we could find out.

Questions therefore maybe left:
- advancedsetting needed (could come later)
- changing default from 4 to 5?
- Anything else?

**Comment by fuzzard** [comment]
I would have leant towards at least advanced setting. I know some users already use advanced settings to make use of max res artwork via imageres/fanartres. I have a feeling users who do such things may prefer the higher quality output.

I'm personally not bothered either way, so whatever you feel like doing is fine for now. As you say, an advanced setting can always be done later if users kick up a fuss I guess

**Comment by da-anda** [comment]
so I did some tests with red images, and q4 really should be the lowest we go for a default. We have quite some vocal users about picture quality IIRC (weren't those the reason we added mipmapping at some point?), and on pictures with flat colored areas or linear gradients, you can really notice added noise to those otherwise smooth areas if you look for it, which enthusiasts will.

An advanced setting would ofc be great, to give those enthusiasts the option to go with higher quality.

**Comment by fritsch** [comment]
I changed the wiki here: https://kodi.wiki/view/Advancedsettings.xml#Library_artwork - can altered / changed / enhanced whatever is wanted.

**Comment by da-anda** [comment]
oh, I didn't add the label to tell you that you should add it, just so that we don't forget to add it to the wiki once the PR is merged. But thanks very much for doing it right away and thanks for adding the advanced setting.

**Comment by fritsch** [comment]
jenkins build this please

File: xbmc/settings/AdvancedSettings.cpp
Comment (enen92) [inline xbmc/settings/AdvancedSettings.cpp:1090]: Mind changing thumb to art or image? Might look peeky but thumb/thumbnail has a long history in Kodi :)
By now thumb is just another generic art type

**Comment by fritsch** [comment]
Done - wiki updated. I lack the time to runtime test it again after the change, so final feedback appreciated.

**Review by enen92 — APPROVED** [review]
Thanks

**Comment by fritsch** [comment]
Was approved - if someone decides differently: Revert is designed into git :-)

**Comment by Glandos** [comment]
Hi,
I discovered this reading https://kodi.tv/article/kodi-nexus-alpha-2/ so I'm a bit late. The best optimization tool for JPEG I've tested so far is https://github.com/mozilla/mozjpeg/ It optimize Huffman tables without losing any information (ie the image is pixel perfect the original).
Of course, reducing quality is the best approach, and you can test all settings using https://squoosh.app/ including MozJPEG.

**Comment by fritsch** [comment]
Years ago we removed special png, jpeg, etc. libs from kodi, so that we don't have to rebuild all that for every platform. That's why we use ffmpeg for everything that we can.
For the jpeg compression itself - it is the same everywhere - reason for this is, that while the DCT is lossless, but the compression only tells which entries in the final lookup table to ignore. 
For Huffman you have a point, as this can win quite something on top.
Though keep in mind, that kodi does not optimize "on picture" but "in general", while there are some images where 14 would still be "optimal" other input would already have issues when compression larger than 5 would be used.

**Comment by da-anda** [comment]
I based my reply on your test image, that clearly shows the block artifacts
in the mentioned area. I also speak from experience working a lot with jpeg
files over the last 10-15 years.
As said, q4 would still be fine with me based on your images, but quality
wise Q3 looks better with read areas. When there are no large red areas, q5
would likely be fine as well.

Peter Frühberger ***@***.***> schrieb am Do., 19. Mai 2022,
15:10:

**Pull request merged**
